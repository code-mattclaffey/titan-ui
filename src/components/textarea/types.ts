import { GenericComponent } from '../../module';

export interface TextAreaInterface extends GenericComponent {
    id: string;
    name: string;
    defaultValue?: string;
    placeholder?: string;
    required?: boolean;
    labelText: string;
}
