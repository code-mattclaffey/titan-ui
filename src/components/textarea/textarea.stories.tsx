import React from 'react';

import { TextArea } from './textarea';

export default {
    title: 'TextArea',
    componentSubtitle: '',
};

export const textTextArea = () => (
    <TextArea
        name="hello-test"
        id="hello-test"
        placeholder="Test placeholder"
        labelText="Hello Test"
    />
);

export const invalidTextArea = () => (
    <TextArea
        name="hello-test"
        id="hello-test"
        labelText="Hello Test"
    />
);

export const validTextArea = () => (
    <TextArea
        name="hello-test"
        id="hello-test"
        defaultValue="Default Text"
        labelText="Hello Test"
    />
);
