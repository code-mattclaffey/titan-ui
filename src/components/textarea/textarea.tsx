import React from 'react';
import classNames from 'classnames';
import { TextAreaInterface } from './types';
import { Label } from '../label/label';

export const TextArea = ({
    additionalClassNames = '',
    name,
    id,
    required = true,
    defaultValue = '',
    labelText = '',
    ...rest
}: TextAreaInterface) => {
    const emptyValue = defaultValue.trim().length === 0;

    const classes = classNames('e-textarea', additionalClassNames);

    const textAreaClasses = classNames('e-input', {
        'e-input--empty': emptyValue,
    });

    return (
        <Label
            additionalClassNames={classes}
            labelText={labelText}
            required={required}
            htmlFor={id}
        >
            <textarea
                id={id}
                name={name}
                className={textAreaClasses}
                data-testid="e-textarea"
                required={required}
                defaultValue={emptyValue ? null : defaultValue}
                {...rest}
            />
        </Label>
    );
};
