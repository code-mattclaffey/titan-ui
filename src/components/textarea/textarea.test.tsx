import React from 'react';

import { render, cleanup } from '@testing-library/react';
import { TextArea } from './textarea';

afterEach(() => {
    cleanup();
});

describe('<TextArea />', () => {
    it('should match snapshot', () => {
        const { container } = render(
            <TextArea
                name="hello-test"
                id="hello-test"
                labelText="Hello World"
                required={false}
            />,
        );

        expect(container).toMatchSnapshot();
    });

    it('should have required attribute to it', () => {
        const { queryByTestId } = render(
            <TextArea
                name="hello-test"
                id="hello-test"
                labelText="Hello World"
                required
            />,
        );

        expect(
            queryByTestId('e-textarea').hasAttribute('required'),
        ).toBeTruthy();
    });

    it('should have placeholder', () => {
        const expectedText = 'This is some placeholder text';
        const { queryByTestId } = render(
            <TextArea
                name="hello-test"
                id="hello-test"
                labelText="Hello World"
                placeholder={expectedText}
            />,
        );

        expect(
            queryByTestId('e-textarea').hasAttribute('placeholder'),
        ).toBeTruthy();
        expect(
            queryByTestId('e-textarea').getAttribute('placeholder'),
        ).toEqual(expectedText);
    });

    it('should have empty class when default is empty', () => {
        const { queryByTestId } = render(
            <TextArea
                name="hello-test"
                id="hello-test"
                labelText="Hello World"
                defaultValue="          "
            />,
        );

        expect(
            queryByTestId('e-textarea').getAttribute('value'),
        ).toBeNull();
        expect(
            queryByTestId('e-textarea').classList.contains(
                'e-input--empty',
            ),
        ).toBeTruthy();
    });

    it('should not have empty class ', () => {
        const expectedText = 'This is some default value text';
        const { getByDisplayValue, queryByTestId } = render(
            <TextArea
                name="hello-test"
                id="hello-test"
                labelText="Hello World"
                defaultValue={expectedText}
            />,
        );

        expect(getByDisplayValue(expectedText)).not.toBeNull();
        expect(
            queryByTestId('e-textarea').classList.contains(
                'e-input--empty',
            ),
        ).toBeFalsy();
    });
});
