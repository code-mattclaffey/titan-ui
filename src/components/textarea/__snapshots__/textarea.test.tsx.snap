// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextArea /> should match snapshot 1`] = `
<div>
  <label
    class="e-label e-textarea"
    data-testid="e-label"
    for="hello-test"
  >
    <span
      class="e-label__text"
    >
      Hello World
    </span>
    <textarea
      class="e-input e-input--empty"
      data-testid="e-textarea"
      id="hello-test"
      name="hello-test"
    />
  </label>
</div>
`;
